### PR TITLE
Move special characters to a dialog.

### LINKED ISSUE #16501: Introduce dialog for special characters.
## 📝 Provide a description of the new feature

Change the special characters plugin to use a dialog instead of a dropdown. Add a menu bar button to use the feature from menu.

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by mremiszewski**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/contributing/git-commit-message-convention.html))

Feature (special-characters): Added menu bar integration to special characters. Closes #16501.

Other (special-characters):  Special characters plugin now uses a dialog (instead of toolbar dropdown).

MINOR BREAKING CHANGE (special-characters):   Special characters plugin now uses a dialog (instead of toolbar dropdown).

---

### Additional information

_For example – encountered issues, assumptions you had to make, other affected tickets, etc._

### CODE REVIEW

**Review by scofalik — CHANGES_REQUESTED** [review]
Other than what was commented:

*   File `specialcharactersnavigationview.ts` should be removed.
*   I don't think we should set tooltip for dropdowns (cc @oleq), we already have text both for label and value.
*   We should try to work on styles a bit when the grid is scrollable, it would be better if the dropdown had margin, so there's space between dropdown and the grid.
*   Q: Wouldn't it be better to refocus editor after click on the special char? This way we would see a caret blinking in the editor, right now I am a bit confused/lost. cc @oleq
*   Q: Wouldn't it be better if we focused first character in the grid instead of categories dropdown? cc @oleq.
*   The toolbar/menu bar component should be "on" when the dialog is opened - we have it like that for AI Assistant and Find and Replace.
*   Fill in PR message.
*   Create an issue for this PR so we can add "Closes" entry to make the changelog better.

Video for reference:

https://github.com/ckeditor/ckeditor5/assets/1502228/23d69ce7-d5c2-40fd-a0b8-01d94ddc422d

File: packages/ckeditor5-special-characters/src/ui/specialcharacterscategoriesview.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	 * TODO
+	 */
+	public dropdownView: LabeledFieldView<DropdownView>;
+
+	/**
+	 * Creates an instance of the {@link module:special-characters/ui/specialcharacterscategoriesview~SpecialCharactersNavigationView}
+	 * class.
+	 *
+	 * @param {module:utils/locale~Locale} locale The localization services instance.
+	 * @param {Iterable.<String>} groupNames The names of the character groups.
```
Comment (scofalik) [inline packages/ckeditor5-special-characters/src/ui/specialcharacterscategoriesview.ts:48]: I don't think we set members types in param anymore.

**Comment by oleq** [comment]
> Other than what was commented:
> 
> * I don't think we should set tooltip for dropdowns (cc @oleq), we already have text both for label and value.

The tooltip shows up because of the initial focus. This is not the only place that acts this way. We may want to rethink it ecosystem-wise but for now I'd just change the label to something meaningful like "Character category" instead of "All" or "Text"

https://github.com/ckeditor/ckeditor5/assets/1099479/88454b24-55f7-4f92-86e7-8cbf0818ca1a

> * We should try to work on styles a bit when the grid is scrollable, it would be better if the dropdown had margin, so there's space between dropdown and the grid.

I'd say this works for me but YMMV

```css
.ck.ck-special-characters > .ck-dialog__content > div > .ck-character-categories {
    padding: var(--ck-spacing-medium) var(--ck-spacing-large);
}
```

image

> * Q: Wouldn't it be better to refocus editor after click on the special char? This way we would see a caret blinking in the editor, right now I am a bit confused/lost. cc @oleq

I briefly checked other editors on the market and those that keep the UI open after insertion do not change the context of the focus. And I got to say it feels natural to me: either you close the UI and change focus context or you keep the UI and preserve focus where clicked.

 OTOH I agree with you that the lack of the caret feels odd. I think we could work around this the way the link feature does using a fake caret (MVP2?). IIRC somebody has already worked on making this fake caret a re-usable piece of code. I'm not sure whether it went to the production, though.

image


> * Q: Wouldn't it be better if we focused first character in the grid instead of categories dropdown? cc @oleq.

That would solve the problem of the tooltip on a drop-down 😛 But I think it's fine the way it is. From the screen user's perspective, the focus order is logical: from the generic (category) to the specific (glyph). I checked other editors on the market and they follow suit. 

> * The toolbar/menu bar component should be "on" when the dialog is opened - we have it like that for AI Assistant and Find and Replace.

👍

**Comment by scofalik** [comment]
Thanks @oleq. Yes, fake caret would be ideal but I figured that's more work. We can do it in next step then.

I understand why the tooltip shows, my question is why we even want to have a tooltip if both information -> field label and field value are displayed anyway. I simply think it is redundant and annoying when it shows. And I agree we should change it everywhere. We can treat it as a next step too.

File: packages/ckeditor5-special-characters/src/ui/specialcharacterscategoriesview.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			} );
+
+			items.add( item );
+		}
+
+		const t = this.locale!.t;
+		const accessibleLabel = t( 'Category' );
+
+		this.dropdownView.label = accessibleLabel;
+		this.dropdownView.bind( 'isEmpty' ).to( this, 'currentGroupName', value => !value );
```
[the remark this answers is not in the document]
Comment (mremiszewski) [reply packages/ckeditor5-special-characters/src/ui/specialcharacterscategoriesview.ts:118]: You are right, we just need to set `isEmpty = false`.

**Comment by Reinmar** [comment]
> * I don't think we should set tooltip for dropdowns (cc @oleq), we already have text both for label and value.

I don't understand why we have tooltips for those dropdowns, when the tooltip repeats what's already visible anyway. Not "why they are displayed" but rather: "why there are there at all".

**Pull request merged**